### ISSUE TITLE
@Nested on Property where the type contains @Input* should consider implicit task dependencies

### ISSUE THREAD

**Issue by lacasseio** [gradle/gradle#13590]
### Expected Behavior

Consider the following:
```
class Foo {
    @Input
    URL url
}

abstract class FooTask {
    @Nested
    abstract ListProperty<Foo> foos
}

def urlGeneratorTask = tasks.register(...) {...}

tasks.register('foo', FooTask) {
    foos.addAll(urlGeneratorTask.flatMap { it.outputProviderOfUrls }.map { new Foo(it) })
}
```
Executing `./gradlew foo` should automatically depend on `urlGeneratorTask` as the provider contains such information and the intent is to use as input for the task `foo`.

### Current Behavior

Implicit task dependencies are not considered and we need to work around this by adding `dependsOn(foos)` inside the constructor of `FooTask`.

### Context

Making everything Provider!

### Steps to Reproduce 

Follow the pseudo code above.

### Your Environment

Build scan URL: None.

**Comment by eskatos** [comment]
Found a related issue https://github.com/gradle/gradle/issues/13386, specifically this [comment](https://github.com/gradle/gradle/issues/13386#issuecomment-642462687).

**Comment by wolfs** [comment]
@lacasseio I looked at your pseudo-code and don't understand why `urlGeneratorTask.flatMap { it.outputProviderOfUrls }` would still depend on `urlGeneratorTask`: `flatMap` loses the dependency information, right? And `outputProviderOfUrls` seems to be a collection of URLs, so it can't be marked with `@Output*`, since Gradle doesn't support non-file outputs.

Though I see the bug you are hinting at. Is this the same as #13386, or do you have a different reproducer?

**Comment by lacasseio** [comment]
That may be the same issues, I don't really have another reproducer. After 2 years, I'm confident I found a workaround to the issue. Still, providers have a lot of gaps making it extremely hard for normal users to properly use them.

**Comment by twukus** [comment]
**Workaround:**
Add a dummy input property to the consumer task that consumes the nested property. 
The value of the property is irrelevant, e.g. empty string.

In the following example I registered the input property in the consumer task constructor:

```
def producer = tasks.register('producer', Producer)

tasks.register('consumer', Consumer) {
    messages = producer.flatMap { it.outputFiles }.map { it.collect { new Message(it.text) } }
}

abstract class Producer extends DefaultTask {
    @OutputFiles
    abstract ListProperty<File> getOutputFiles()
}

abstract class Consumer extends DefaultTask {
    Consumer() {
        getInputs().property("messages-ValueProducer-Receiver", messages.map { "" });
    }

    @Nested
    abstract ListProperty<Message> getMessages()
}

class Message {
    @Input
    String text
}
```

Test run:
```
$ gradle consumer --dry-run
:producer SKIPPED
:consumer SKIPPED

BUILD SUCCESSFUL in 488ms
```

**Comment by CRogers** [comment]
Hit this recently and left me stumped for ages why it wasn't working. This is legitimately a bug imo - Gradle should consider the implicit task dependencies from the provider that is set on the `@Nested` property

**Comment by CRogers** [comment]
Here is a minimal reproducer: https://github.com/CRogers/gradle-task-dependencies-lost-on-nested

The actual case I encountered involves setting a `@Nested SetProperty<NestedObject>` from an artifact view, so we can't just create the nested object, set the nested properties from the original source so they get the dependencies (as we do not know how many objects to make on the set property.).

**Comment by octylFractal** [comment]
@CRogers Your reproducer doesn't quite make sense to me, as a non-dry-run doesn't succeed on 8.14 due to pulling on the output value of the task too early. After further modifications to fix various issues, I arrived at this `build.gradle`: https://gist.github.com/octylFractal/32f622117876626f31cda5e0eee087ff
You can see the changes I made at: https://gist.github.com/octylFractal/32f622117876626f31cda5e0eee087ff/revisions

This works in 8.14. Does that solve your issue? If so, I'm inclined to close this as there appears to be no bug in the current version of Gradle. It appears that `@Nested` is tracking input task dependencies now.

**Comment by Vampire** [comment]
I agree with @twukus, in the modified version by @octylFractal if you switch back `map` to `flatMap` it should work but does not.
If you add a manual `dependsOn` from `someTask` to `otherTask` or add the work-around mentioned above, i.e. adding 
```groovy
SomeTask() {
    inputs.property('work-around', something.map { '' })
}
```
it works properly.

**Comment by CRogers** [comment]
+1 to the previous two comments, but @twukus the error that appears when you do not dry run on my minimal reproducer is:

```
Caused by: org.gradle.api.InvalidUserCodeException: Querying the mapped value of flatmap(provider(task 'otherTask', class OtherTask)) before task ':otherTask' has completed is not supported
```

...which is exactly caused by this bug - `otherTask` did not get run before `someTask` as Gradle did not make the task dependency. It works find if you have a manual dependsOn, which is what we want to avoid.

I'm pretty sure, testing with `--dry-run` should be enough to verify the task dependencies, fwiw.

**Comment by octylFractal** [comment]
Ah, I thought that exception was happening earlier than execution. I adjusted my reproducer to use `flatMap`, and it should serve as a better reproducer due to properly wiring everything for if this issue is fixed, so it won't have potential sources of confusion.
